File: OrderProcessor/OrderProcessor/Services/OrderValidator.cs
using OrderProcessor.Models;

namespace OrderProcessor.Services;

public sealed class OrderValidator
{
    public void Validate(Order order)
    {
        ArgumentNullException.ThrowIfNull(order, nameof(order));

        if (order.Items is null || !order.Items.Any())
            throw new InvalidOperationException("Order must contain at least one item.");
        
        foreach (var item in order.Items)
        {
            if (item.Quantity <= 0)
                throw new InvalidOperationException($"Item '{item.Name}' must have a quantity greater than zero.");
            
            if (item.Price < 0)
                throw new InvalidOperationException($"Item '{item.Name}' cannot have a negative price.");
        }
    }
}

File: OrderProcessorCpp/src/services/OrderValidator.cpp
#include "OrderValidator.h"

bool OrderValidator::Validate(const Order& order)
{
    if (order.orderId.empty()) return false;
    if (order.items.empty()) return false;
    
    for (auto const& item : order.items)
    {
        if (item.quantity <= 0 || item.price < 0.0) return false;
    }
    
    return true;
}
